userspace/libscap/ringbuffer/ringbuffer.c

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright (C) 2022 The Falco Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+#include <stdint.h>
+#include <stdio.h>
+#include <math.h>
+#include <scap.h>
+
+#define MAX_ALLOWED_VALUES 12
+
+/* Dimension of a single per-CPU buffer. */
+unsigned long per_cpu_buffer_dim;
+
+void set_per_cpu_buffer_dim(unsigned long buf_dim)
+{
+	per_cpu_buffer_dim = buf_dim;
+}
+
+int32_t check_per_cpu_buffer_num_pages(char* last_err, unsigned long buf_num_pages)
+{
+	/* Allowed page numbers are all the power of 2 from 128(2^7) pages to 256Kpages(2^18) */
+	int allowed_num_pages_values[MAX_ALLOWED_VALUES] = {pow(2, 7), pow(2, 8), pow(2, 9), pow(2, 10), pow(2, 11), pow(2, 12), pow(2, 13), pow(2, 14), pow(2, 15), pow(2, 16), pow(2, 17), pow(2, 18)};
```

Comment:
makes sense, ty! for what concerns the shared header, in the end, these are just 4 lines

```
if(dim == 0 ||
	 dim % PAGE_SIZE != 0 ||
	 (dim & (dim - 1)) != 0 ||
	 dim < 128 * PAGE_SIZE)
{
pr_err("the specified per-CPU ring buffer dimension (%lu) is not allowed! Please use a power of 2 and a multiple of the actual page_size (%lu)!\n", dim, PAGE_SIZE);
return -EINVAL;
}
```

 probably i would avoid the shared helper :thinking:

Reply:
Oh yeah i've added this `dim < 128 * PAGE_SIZE` just to avoid stupid value with sysctl, but also `dim < 2 * PAGE_SIZE` it's ok, this is the equivalent of this check `sizeof(allowed_num_pages_values) / sizeof(allowed_num_pages_values[0])`.  The problem with unyfing the kernel side with the userspace side is that we are checking 2 different things: 

* userspace side we are checking that the number of pages is a power of 2 (in a nutshell we are validating the input param from the `scap_open_args`) ->  the idea of this method is to become part of  a new possible `v-table` method, `validate_engine_params` that allow us to validate all the params before opening the engine with the init method :thinking: 
* kernel side we are checking directly the dimension in bytes of the single buffer

Do we really want to unify these checks?

Reply:
Ah, here is where the pages vs bytes confusion comes in :D 

This is probably overkill but what about `bool is_valid_ringbuf_size(size_t count, size_t unit)` which would validate that:
* count * unit does not overflow size_t
* count * unit is a valid ring buffer size

userspace would call it with unit=PAGE_SIZE, the kernel would call it with unit=1

...as long as we *really* want userspace to deal with page-sized values; how about we push bytes across the whole API and leave pages (if you prefer them) in the user-facing parts only (command line options?)

Reply:
I was thinking about the same thing, probably it is simpler to use the page size in all the APIs and leave the page number for the command line. The remaining question here is, how do we want to validate the buffer dimension?
* a fixed array with 10/12 dimension
* or every power of 2 is good?

Reply:
@FedeDP @leogr @incertum

Reply:
@gnosek, please let us know if you agree with that :point_up: , since we plan to merge this PR this week and then proceed with a code freeze starting next week. 

If we don't find a consensus around this PR, it would be an issue that will likely delay the Falco release :crying_cat_face: . So let us know if there's any other blocking issue in your opinion. :pray: 

Thank you.

Reply:
@leogr, I'm not sure it gets much simpler than https://man7.org/linux/man-pages/man2/getpagesize.2.html :)

I'd prefer to see both userspace and kernel use the same code to check the buffer size but I'm not going to block the release over it (and over the module param vs ioctl)

Reply:
> @leogr, I'm not sure it gets much simpler than https://man7.org/linux/man-pages/man2/getpagesize.2.html :)

My proposal was just a wrapper around it, in case there were edge cases on some arch. But it's ok for me to do not to provide a dummy API :smile_cat:  
 
> I'd prefer to see both userspace and kernel use the same code to check the buffer size but I'm not going to block the release over it (and over the module param vs ioctl)

If I understood correctly, we would have like four lines of code replicated in several points. But it will still be the same logic (@Andreagit97, right?)
If we get the same behavior on the userspace and kernel, it is ok for me just to replicate the same logic and then improve the implementation in subsequent PRs (since this is already big enough, IMHO).

Reply:
Few comments, hopefully useful, if not disregard. @leogr probably best to make a final call before asking @Andreagit97 for another refactor, so that Andrea can do one final round of commits that will be accepted. That way we would be able to land this for 0.33.

> I believe it makes sense to accept bytes as units 

What puzzles me is that buffer size calculation before was based on hard-coded multiple of PAGE_SIZE etc and it was fine (at least for bpf, never looked to deep into kmod). Also before hard-coded values were always trusted, no validations. Initial idea was to just make this a parameter. There is quite a long new discussion going on, so please pardon me to re-ask questions that were answered before. 

What is the justification that this now needs to be all refactored at once, aka multiple of PAGE_SIZE is not ok anymore? Why do we now need overly abundant checks versus none before? End user input checks would have satisfied the state before the refactor, but all checks make sense and I am always in favor of more checks, just need to find a balance for this initial refactor?

Perhaps there are really good justifications, in that case, would it make sense to go with an initial refactor to land the 0.33 release and polish it more for 0.34?

> and then validate the following simple rules:
> 
> * must be a power of 2
> * must be a multiple of PAGE_SIZE
> * must be greater than 2*PAGE_SIZE
> 
> I wouldn't introduce any upper limit at this level. To help consumers do the proper math, I will also provide a simple API to get the PAGE_SIZE value.
> 

@leogr @Andreagit97 re no upper limit this is fine. 
[Not sure under what conditions you would go higher than 2^18 * PAGE_SIZE, I think highest you realistically need is 2^12 or 2^13 in heavy production environments. However, please note that those statements are based on Falco use case only (simple consumer - only interesting system calls) and x86_64. either approach will be fine as we have the warning message when memory cannot be allocated. The error message also contains a hint that you should reduce the buffer dimension.]

@leogr Re end users needing to first get the PAGE_SIZE value, and then needing to do the math, would vote against. We had that [initial discussion](https://github.com/falcosecurity/falco/pull/2164#issuecomment-1220479171) and everyone thought multiples of PAGE_SIZE was best and simplest end user experience, what changed?
Perhaps we should do an end user poll in Slack to learn what more end users would prefer to avoid making wrong decisions re what we think is best end user experience?

If it's about needing some unit=1 kernel side I think we can implement this as math in the code to suit the respective validation check use case. As we have PAGE_SIZE available just like before.

> PS If necessary, I'd leave the fixed array solution to the consumer, but not here.

@leogr @Andreagit97 for Falco yes I think this would be simplest end user experience.

Reply:
I answered here https://github.com/falcosecurity/libs/pull/584#issuecomment-1246407597 :)

Reply:
I was in a rush when commenting, so perhaps I did not well explain my thoughts :sweat_smile:  Let me clarify 

> Few comments, hopefully useful, if not disregard. @leogr probably best to make a final call before asking @Andreagit97 for another refactor, so that Andrea can do one final round of commits that will be accepted. That way we would be able to land this for 0.33. :

@Andreagit97 and I had a quick chat before my comment, and the refactoring effort seemed acceptable, so I did my proposal.

> What is the justification that this now needs to be all refactored at once, aka multiple of PAGE_SIZE is not ok anymore? Why do we now need overly abundant checks versus none before? End user input checks would have satisfied the state before the refactor, but all checks make sense and I am always in favor of more checks, just need to find a balance for this initial refactor?

> Perhaps there are really good justifications, in that case, would it make sense to go with an initial refactor to land the 0.33 release and polish it more for 0.34?

Our new rationale is:  PAGE_SIZE is still ok, but not at this API level (and in this PR). 

After @gnosek raised his points, we realized that using only PAGE_SIZE would restrict us from future implementations that may not use the same use (btw, for the kmod is already so, it does not actually need the PAGE_SIZE alignment).
Using a byte as a unit and doing the proper validation will allow us to introduce new cases without breaking the backward compatibility.

I used the term "consumer" to indicate the libs' consumers (e.g., Falco).  For example, implementing the fixed array logic (with a set of precalculated values based on the PAGE_SIZE) directly in Falco is extremely easy. Alternatively,  sinsp could provide a helper function to do the math, but this can be quickly introduced with another PR any time later, even after the release of Falco 0.33. (Moving some logic down from Falco to sinsp is generally straightforward. The vice-versa is not always the case.)

I agree that end users must not work with bytes or do the math by themself. So, in the end, my suggestion was: to use bytes in libs with proper validation and just move the easy PAGE_SIZE multiplication in Falco.


> > I wouldn't introduce any upper limit at this level. To help consumers do the proper math, I will also provide a simple API to get the PAGE_SIZE value.
> 
> @leogr @Andreagit97 re no upper limit this is fine. [Not sure under what conditions you would go higher than 2^18 * PAGE_SIZE, I think highest you realistically need is 2^12 or 2^13 in heavy production environments. However, please note that those statements are based on Falco use case only (simple consumer - only interesting system calls) and x86_64. either approach will be fine as we have the warning message when memory cannot be allocated. The error message also contains a hint that you should reduce the buffer dimension.]

I don't have a strong opinion on that. I thought that having an upper limit could reduce the freedom to experiment, and at the same time, I don't see a not opinionated way to calculate the upper limit. 
For me, it's OK, either with or without the upper limit.
It's up to Andrea to decide. He perhaps has more thoughts on this.


> @leogr Re end users needing to first get the PAGE_SIZE value, and then needing to do the math, would vote against. We had that [initial discussion](https://github.com/falcosecurity/falco/pull/2164#issuecomment-1220479171) and everyone thought multiples of PAGE_SIZE was best and simplest end user experience, what changed? Perhaps we should do an end user poll in Slack to learn what more end users would prefer to avoid making wrong decisions re what we think is best end user experience?
> 
> If it's about needing some unit=1 kernel side I think we can implement this as math in the code to suit the respective validation check use case. As we have PAGE_SIZE available just like before.
> 
> > PS If necessary, I'd leave the fixed array solution to the consumer, but not here.
> 
> @leogr @Andreagit97 for Falco yes I think this would be simplest end user experience.

I don't think we changed the outcome of the initial discussion. Above, I tried to clarify that my suggestion was only about how to achieve the goal and not about changing the plan. 

Let's unblock this PR this week so that at least we will have the feature in Falco. The poll is a good idea, but perhaps later with no rush. Maybe after users have tried the feature, they will also have more feedback to provide.

To wrap up, my proposal:
 - use bytes in libs (this PR)
 - do proper input validation at all levels (this PR)
 - upper limit (this or another PR,  let Andrea decide)
 - unblock this PR this week
 - implement the fixed array solution in Falco (another PR) - I still think it is the best UX
 - maybe in the future (in another PR), provide some help to make the previous point easier to implement 


Hoping this helped to clarify, and I apologize for my previous comments that perhaps were misleading :pray:

Reply:
> To wrap up, my proposal:
> 
>    *  use bytes in libs (this PR)
>    * do proper input validation at all levels (this PR)
>    * upper limit (this or another PR, let Andrea decide)
>    * unblock this PR this week
>    * implement the fixed array solution in Falco (another PR) - I still think it is the best UX
>    * maybe in the future (in another PR), provide some help to make the previous point easier to implement
> 
> Hoping this helped to clarify, and I apologize for my previous comments that perhaps were misleading pray

Agree with @leogr on all these points (I've chosen to not implement any upper bound, we will add it if necessary).

One thing that we can discuss but regards Falco and not this PR is the UX. I was thinking if a fixed array of power of 2 is better than a fixed array of pages numbers, as I proposed here:

> BTW also passing directly a power of 2 seems a smart idea (for example you want an 8 MByte buffer, you can simply pass 23 to Falco so 1<<23), so again maybe an array that checks a set of powers of 2, @gnosek proposed it for the kernel module param but we can reuse the idea for the user interface in Falco

If you have any suggestions on it please feel free to comment :)

Reply:
Thanks @leogr ❤️ for clarifying, it totally makes sense now. In summary, this sounds like an excellent plan. 

Re no upper limit yeah just remove it / leave it as is (no upper limit) as the error message will take care of it when memory cannot be allocated.

Re "fixed array" for Falco, agree seems simplest for Falco.

Thanks @Andreagit97 for all the hard work, it ended up being a larger refactor than at least I initially anticipated.